src/sync/cond.go

```diff
@@ -0,0 +1,80 @@
+package sync
+
+import "internal/task"
+
+type Cond struct {
+	L Locker
+
+	unlocking *earlySignal
+	blocked   task.Stack
+}
+
+// earlySignal is a type used to implement a stack for signalling waiters while they are unlocking.
+type earlySignal struct {
+	next *earlySignal
+
+	signaled bool
+}
+
+func (c *Cond) trySignal() bool {
+	// Pop a blocked task off of the stack, and schedule it if applicable.
+	t := c.blocked.Pop()
+	if t != nil {
+		scheduleTask(t)
+		return true
+	}
+
+	// If there any tasks which are currently unlocking, signal one.
+	if c.unlocking != nil {
+		c.unlocking.signaled = true
+		c.unlocking = c.unlocking.next
+		return true
+	}
```

Comment:
So this code removes one entry from the `unlocking` queue. At the same time, `Wait` appears to pop off the earlySignal frame unconditionally. So wouldn't that result in a conflict (likely resulting in a nil pointer dereference in `Wait`)?

Reply:
There is no conflict, as marking it as signaled skips the pop.